core/templates/dev/head/components/state-editor/state-editor-properties-services/state-property.service.ts

```diff
@@ -15,75 +15,85 @@
+  // property. THIS MUST BE SPECIFIED BY SUBCLASSES.
+  setterMethodKey: string;
+  displayed: any;
+  stateName: string;
+  savedMemento: any;
 
-        // The name of the state.
-        this.stateName = stateName;
-        // The current value of the property (which may not have been saved to
-        // the frontend yet). In general, this will be bound directly to the UI.
-        this.displayed = angular.copy(value);
-        // The previous (saved-in-the-frontend) value of the property. Here,
-        // 'saved' means that this is the latest value of the property as
-        // determined by the frontend change list.
-        this.savedMemento = angular.copy(value);
-      },
-      // Returns whether the current value has changed from the memento.
-      hasChanged: function() {
-        return !angular.equals(this.savedMemento, this.displayed);
-      },
-      // The name of the setter method in ExplorationStatesService for this
-      // property. THIS MUST BE SPECIFIED BY SUBCLASSES.
-      setterMethodKey: null,
-      // Transforms the given value into a normalized form. THIS CAN BE
-      // OVERRIDDEN BY SUBCLASSES. The default behavior is to do nothing.
-      _normalize: function(value) {
-        return value;
-      },
-      // Validates the given value and returns a boolean stating whether it
-      // is valid or not. THIS CAN BE OVERRIDDEN BY SUBCLASSES. The default
-      // behavior is to always return true.
-      _isValid: function(value) {
-        return true;
-      },
-      // Updates the memento to the displayed value.
-      saveDisplayedValue: function() {
-        if (this.setterMethodKey === null) {
-          throw 'State property setter method key cannot be null.';
-        }
+  constructor(private alertsService: AlertsService) {
+    this.setterMethodKey = null;
+  }
 
-        this.displayed = this._normalize(this.displayed);
-        if (!this._isValid(this.displayed) || !this.hasChanged()) {
-          this.restoreFromMemento();
-          return;
-        }
+  init(stateName: string, value: any): void {
+    if (this.setterMethodKey === null) {
+      throw 'State property setter method key cannot be null.';
+    }
+    // The name of the state.
+    this.stateName = stateName;
+    // The current value of the property (which may not have been saved to
+    // the frontend yet). In general, this will be bound directly to the UI.
+    this.displayed = angular.copy(value);
+    // The previous (saved-in-the-frontend) value of the property. Here,
+    // 'saved' means that this is the latest value of the property as
+    // determined by the frontend change list.
+    this.savedMemento = angular.copy(value);
+  }
 
-        if (angular.equals(this.displayed, this.savedMemento)) {
-          return;
-        }
+  // Returns whether the current value has changed from the memento.
+  hasChanged(): boolean {
+    return !angular.equals(this.savedMemento, this.displayed);
```

Comment:
angular.equals should also be replaced.

You can create this function to check if these two are equivalent:
```
private isEquivalent(a: any, b: any): boolean {
    // Create arrays of property names
    var aProps = Object.getOwnPropertyNames(a);
    var bProps = Object.getOwnPropertyNames(b);
    if (aProps.length !== bProps.length) {
      return false;
    }
    for (var i = 0; i < aProps.length; i++) {
      var propName = aProps[i];
      if (a[propName] !== b[propName]) {
        return false;
      }
    }
    return true;
  }
```

Replace the angular.equals here and everywhere.

Reply:
Done.